### ISSUE TITLE
CI: WML unit tests results make it very hard to find errors

### ISSUE THREAD

**Issue by stevecotton** [wesnoth/wesnoth#5616]
@CelticMinstrel the WML unit test script's results are currently very confusing. Below is part of the output of the c611128 validation (which passes, none of these are problems). However, this makes it very hard to work out why the CI failed on other runs - there's one other error, somewhere.

```
Running test test_move_fail_6
Error (strict mode, strict_level = 1): wesnoth reported on channel warning replay
20210314 23:50:58 warning replay: Warning: Path data contained something which could not be parsed to a sequence of locations:
 config = x = 16,15,14,13,12,11
y = 3,3,3,3,3,bock

FAIL TEST (BROKE STRICT): test_move_fail_6

Running test test_store_unit_defense_deprecated
Error (strict mode, strict_level = 1): wesnoth reported on channel error deprecation
20210314 23:50:59 error deprecation: [store_unit_defense] has been deprecated and will be removed in version 1.17.0.
  This function returns the chance to be hit, high values represent bad defenses. Using [store_unit_defense_on] is recommended instead.
20210314 23:50:59 error deprecation: [store_unit_defense] has been deprecated and will be removed in version 1.17.0.
  This function returns the chance to be hit, high values represent bad defenses. Using [store_unit_defense_on] is recommended instead.
20210314 23:50:59 error deprecation: [store_unit_defense] has been deprecated and will be removed in version 1.17.0.
  This function returns the chance to be hit, high values represent bad defenses. Using [store_unit_defense_on] is recommended instead.
20210314 23:50:59 error deprecation: [store_unit_defense] has been deprecated and will be removed in version 1.17.0.
  This function returns the chance to be hit, high values represent bad defenses. Using [store_unit_defense_on] is recommended instead.
20210314 23:50:59 error deprecation: [store_unit_defense] has been deprecated and will be removed in version 1.17.0.
  This function returns the chance to be hit, high values represent bad defenses. Using [store_unit_defense_on] is recommended instead.
20210314 23:50:59 error deprecation: [store_unit_defense] has been deprecated and will be removed in version 1.17.0.
  This function returns the chance to be hit, high values represent bad defenses. Using [store_unit_defense_on] is recommended instead.
FAIL TEST (BROKE STRICT): test_store_unit_defense_deprecated

Running test alice_kills_bob
PASS TEST (VICTORY): alice_kills_bob

Running test bob_kills_alice_on_retal
PASS TEST (VICTORY): bob_kills_alice_on_retal

Running test alice_kills_bob_levelup
PASS TEST (VICTORY): alice_kills_bob_levelup

Running test bob_kills_alice
PASS TEST (VICTORY): bob_kills_alice

Running test alice_kills_bob_on_retal
PASS TEST (VICTORY): alice_kills_bob_on_retal

Running test alice_kills_bob_on_retal_levelup
PASS TEST (VICTORY): alice_kills_bob_on_retal_levelup

Running test test_wml_menu_items_2
Error (strict mode, strict_level = 1): wesnoth reported on channel warning wml
20210314 23:51:08 warning wml: The following conditional test unexpectedly failed:
	[variable]
		boolean_equals=yes
		name="result"
	[/variable]
	
	Interpolated to:
	[variable]
		boolean_equals=yes
		name="result"
	[/variable]
	
	Note: The variable result currently has the value false.
FAIL TEST: test_wml_menu_items_2

Running test filter_formula_unit_error
Error (strict mode, strict_level = 1): wesnoth reported on channel error scripting/lua
20210314 23:51:09 error scripting/lua: Formula error in formula:1
In formula + 
Error: Illegal unary operator: '+'
stack traceback:
	[C]: in ?
	[C]: in field 'eval_conditional'
	lua/wml-flow.lua:17: in local 'cmd'
	lua/wml-utils.lua:144: in field 'handle_event_commands'
	lua/wml-flow.lua:19: in local 'cmd'
	lua/wml-utils.lua:144: in field 'handle_event_commands'
	lua/wml-flow.lua:5: in function <lua/wml-flow.lua:4>
FAIL TEST (BROKE STRICT): filter_formula_unit_error
```

**Comment by Pentarctagon** [comment]
What does it looks like when it does fail?

**Comment by stevecotton** [comment]
If it's another WML test that's causing the failure, it might look like this:

```
PASS TEST: event_handlers_in_events_6
PASS TEST: event_handlers_in_events_7
PASS TEST: event_handlers_in_events_8
Error (strict mode, strict_level = 1): wesnoth reported on channel warning wml
20210315 21:34:47 warning wml: The following conditional test unexpectedly failed:
	[have_unit]
		side=2
		[filter_vision]
			side=40
		[/filter_vision]
	[/have_unit]
	
	Interpolated to:
	[have_unit]
		side=2
		[filter_vision]
			side=40
		[/filter_vision]
	[/have_unit]
	
	
FAIL TEST: filter_vision

Failure, Wesnoth returned 1 FAIL but we expected 0 PASS
Running 141 tests (scatter_units, has_ally, ...
```

But if it's unrelated, it might look like this, somewhere else in a log that's about 2200 lines long:

```
Executing WML indentation check
make: Entering directory '/home/wesnoth-CI/data/tools'
wmlindent: ../../data/core/hotkeys.cfg changed
make: Leaving directory '/home/wesnoth-CI/data/tools'
WML indentation check failed! (checkindent)
```

**Comment by CelticMinstrel** [comment]
In the commit you linked you can scroll down to the end and confirm that 228 of 228 tests passed so there's no need to go searching for failures in the WML tests.

So if I understand correctly, you just want the runner to be more clear about when a failure was expected? Or would you like a list of tests that failed in the summary at the end?

Another thing that would help is if the various checks weren't all in the same giant log, not sure if there's anything we can do about that though…

**Comment by stevecotton** [comment]
> In the commit you linked you can scroll down to the end and confirm that 228 of 228 tests passed so there's no need to go searching for failures in the WML tests.

Dude, do a text search for "shared_ptr" in https://github.com/wesnoth/wesnoth/pull/5607#discussion_r594372252 , and then consider how many lines away that was from the line that you noticed. :grinning:

I want the output to be less likely to trigger my "ooh, there's the error" sense when scrolling through the log.

**Comment by stevecotton** [comment]
@CelticMinstrel I see the confused emoji - have you understood, or do you want me to write another explanation?

**Comment by CelticMinstrel** [comment]
A direct answer to my questions would be nice.

**Comment by stevecotton** [comment]
> So if I understand correctly, you just want the runner to be more clear about when a failure was expected?

When a test passes, I want it to be less verbose and not print something that looks like a test failure. If a test passes, I don't want it to print `FAIL TEST` at all.

> Or would you like a list of tests that failed in the summary at the end?
> In the commit you linked you can scroll down to the end and confirm that 228 of 228 tests passed so there's no need to go searching for failures in the WML tests.

That is of very little help. When a tool has started printing `FAIL TEST` on line 1641 of the log and continued to print what look like failure messages until line 1890 of the log, printing `Result: 228 of 228 tests passed` on line 2076 still leaves the reader with a lot of noise to sort though while searching through the log trying to find what caused the real problem. Over 10% of the lines of the entire log are what looks like error diagnostics from the WML tests; the real problem might be further up the log, above all these `FAIL TEST` lines, so the reader still has to scroll upwards while thinking "all of this is verbose output from a passing test".

To put it another way, who is expected to read the log and benefit from `FAIL TEST` messages being printed by tests that pass?

**Comment by Pentarctagon** [comment]
I'm very tempted to revert https://github.com/wesnoth/wesnoth/commit/9969e1c34acf6da93f0833cfd76e60387af69916 until perhaps a better solution can be found.  As already mentioned this makes the CI logs much harder to read, as well as the logs stating a a test failed when what really happened is that a test *that was expected to fail* returned the FAIL result and as such is actually successful is simply not how these should be displayed.

If a test returns a result that it shouldn't, and therefore fails, then *that test's output alone* should be displayed.

**Comment by CelticMinstrel** [comment]
I don't think that's the right commit to revert if you were going to revert one to make the unit test logs easier to read. All that does is add a two-line summary at the very end of the tests.

**Comment by CelticMinstrel** [comment]
This should be fixed by 1273aba5d86ed0df2486a618d01c49da089dfdf1.

**Comment by CelticMinstrel** [comment]
Although the above commit does address this issue, it seems CI is invoking the tests with `-v`, so they're still outputting all that information. Is there some other reason to pass `-v`?

**Comment by Pentarctagon** [comment]
Not that I'm aware of.  Even if there were previously, it doesn't sound like it'd apply anymore.

**Comment by Pentarctagon** [comment]
The -v is now removed and the output is much better, so closing this.
